### PR TITLE
AKCORE-83: Handle lock timeout on messages in share partition (1/N)

### PR DESCRIPTION
**Pull request by adixitconfluent**
### About
Added the record lock duration timeout config to brokers

### CODE REVIEW

File: core/src/main/java/kafka/server/SharePartitionManager.java [hunk 1]
```diff
@@ -68,19 +71,24 @@ public class SharePartitionManager {
...
+    SharePartitionManager(ReplicaManager replicaManager, Time time, ShareSessionCache cache, Map<SharePartitionKey, SharePartition> partitionCacheMap,
+                          int recordLockDurationMs) {
         this.replicaManager = replicaManager;
         this.time = time;
         this.cache = cache;
         this.partitionCacheMap = partitionCacheMap;
         this.fetchQueue = new ConcurrentLinkedQueue<>();
         this.processFetchQueueLock = new AtomicBoolean(false);
+        this.recordLockDurationMs = recordLockDurationMs;
+        this.timer = new SystemTimerReaper("sharePartitionReaper", new SystemTimer("sharePartitionTimer"));
```
Comment (apoorvmittal10) [inline core/src/main/java/kafka/server/SharePartitionManager.java:91]:
```suggestion
        this.timer = new SystemTimerReaper("share-group-lock-timeout-reaper", new SystemTimer("share-group-lock-timeout"));
```

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 2]
```diff
@@ -814,6 +816,8 @@ object KafkaConfig {
   val ShareGroupHeartbeatIntervalMsDoc = "The heartbeat interval given to the members of a share group."
   val ShareGroupMinHeartbeatIntervalMsDoc = "The minimum heartbeat interval for share group members."
   val ShareGroupMaxHeartbeatIntervalMsDoc = "The maximum heartbeat interval for share group members."
+  val ShareGroupRecordLockDurationMsDoc = "Share-group record acquisition lock duration in milliseconds."
```
Comment (apoorvmittal10) [inline core/src/main/scala/kafka/server/KafkaConfig.scala:819]:
```suggestion
  val ShareGroupRecordLockDurationMsDoc = "The record acquisition lock duration in milliseconds for share group."
```

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 3]
```diff
@@ -814,6 +816,8 @@ object KafkaConfig {
...
+  val ShareGroupMaxRecordLockDurationMsDoc = "Share-group record acquisition lock maximum duration in milliseconds."
```
Comment (apoorvmittal10) [inline core/src/main/scala/kafka/server/KafkaConfig.scala:820]:
```suggestion
  val ShareGroupMaxRecordLockDurationMsDoc = "The record acquisition lock maximum duration in milliseconds for share group."
```

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 4]
```diff
@@ -312,6 +312,8 @@ object KafkaConfig {
   val ShareGroupHeartbeatIntervalMsProp = "group.share.heartbeat.interval.ms"
   val ShareGroupMinHeartbeatIntervalMsProp = "group.share.min.heartbeat.interval.ms"
   val ShareGroupMaxHeartbeatIntervalMsProp = "group.share.max.heartbeat.interval.ms"
+  val ShareGroupRecordLockDurationMsProp = "group.share.record.lock.duration.ms"
+  val ShareGroupMaxRecordLockDurationMsProp = "group.share.record.lock.duration.max.ms"
```
Comment (apoorvmittal10) [inline core/src/main/scala/kafka/server/KafkaConfig.scala:316]: @AndrewJSchofield These seems confusing to me, just re-read the KIP. Can't we have just one config which is `group.share.record.lock.duration.ms` and has minimum and maximum, why do we require 2 of them?

File: core/src/main/scala/kafka/server/BrokerServer.scala [hunk 5]
```diff
@@ -394,7 +394,7 @@ class BrokerServer(
 
       val shareFetchSessionCache : ShareSessionCache = new ShareSessionCache(config.shareGroupMaxGroups * config.shareGroupMaxSize,
         KafkaServer.MIN_INCREMENTAL_FETCH_SESSION_EVICTION_MS)
-      val sharePartitionManager = new SharePartitionManager(replicaManager, Time.SYSTEM, shareFetchSessionCache)
+      val sharePartitionManager = new SharePartitionManager(replicaManager, Time.SYSTEM, shareFetchSessionCache, config.shareGroupRecordLockDurationMs)
```
Comment (apoorvmittal10) [inline core/src/main/scala/kafka/server/BrokerServer.scala:397]: Where do we use `config.shareGroupMaxRecordLockDurationMs`?

File: core/src/main/scala/kafka/server/KafkaConfig.scala (hunk 4 above)
Comment (AndrewJSchofield) [reply core/src/main/scala/kafka/server/KafkaConfig.scala:316]: One is the duration of the record locks for a group which does not have its own configuration for lock duration, and the other is the maximum value that is permitted for the group configuration for lock duration.

Comment (apoorvmittal10) [reply core/src/main/scala/kafka/server/KafkaConfig.scala:316]: Though it must be something you already thought about and I am sure I am missing something. But what about a single configuration which has a min, max and a default value.

Is it more about that `group.share.record.lock.duration.ms` configuration cannot be more that the `group.share.record.lock.max.duration.ms`, I think that makes sense. In that case we do not need to define `group.share.record.lock.duration.ms` in KafkaConfig as that should be part of consumer config, correct?

Comment (apoorvmittal10) [reply core/src/main/scala/kafka/server/KafkaConfig.scala:316]: Yeah, I think it's latter and KIP says that `group.share.record.lock.duration.ms` config is specified by consumer clients.

File: core/src/main/scala/kafka/server/BrokerServer.scala (hunk 5 above)
Comment (adixitconfluent) [reply core/src/main/scala/kafka/server/BrokerServer.scala:397]: It is only used to compare `group.share.record.lock.duration.ms` and `group.share.record.lock.max.duration.ms` [here](https://github.com/confluentinc/kafka/pull/1151/files#diff-cbe6a8b71b05ed22cf09d97591225b588e9fca6caaf95d3b34a43262cfd23aa6R2387)

File: core/src/main/scala/kafka/server/KafkaConfig.scala (hunk 4 above)
Comment (adixitconfluent) [reply core/src/main/scala/kafka/server/KafkaConfig.scala:316]: I am not sure, is there any change expected in the code?

**Review by apoorvmittal10 — APPROVED** [review]
LGTM. We might have to change a bit of config reads when we support dynamic configuration for lock timeout.

**Pull request merged**